src/batched/dense/impl/KokkosBatched_Gemm_Armpl_Impl.hpp

```diff
@@ -0,0 +1,300 @@
+      Kokkos::Impl::throw_runtime_exception(os.str());
+    }
+    delete __Adp;
+    delete __Bdp;
+  }
+
+ public:
+  BatchedArmplGemm(HandleType *const handle, ScalarType alpha, AViewType A,
+                   BViewType B, ScalarType beta, CViewType C)
+      : __handle(handle), __A(A), __B(B), __C(C), __alpha(alpha), __beta(beta) {
+    __ninter = __handle->get_tpl_params()[0];
+
+    if (std::is_same<ArgBatchSzDim, BatchLayout::Left>::value) {
+      __Am     = __A.extent(1);
+      __An     = __A.extent(2);
+      __Bm     = __B.extent(1);
+      __Bn     = __B.extent(2);
+      __Cm     = __C.extent(1);
+      __Cn     = __C.extent(2);
+      __nbatch = __C.extent(0);
+    } else {
+      __Am     = __A.extent(0);
+      __An     = __A.extent(1);
+      __Bm     = __B.extent(0);
+      __Bn     = __B.extent(1);
+      __Cm     = __C.extent(0);
+      __Cn     = __C.extent(1);
+      __nbatch = __C.extent(2);
+    }
+
+    __Ajstrd = __ninter;
+    __Aistrd = __Ajstrd * __An;
+    __Abstrd = __Aistrd * __Am;
+
+    __Bjstrd = __ninter;
+    __Bistrd = __Bjstrd * __Bn;
+    __Bbstrd = __Bistrd * __Bm;
+
+    __Cjstrd = __ninter;
+    __Cistrd = __Cjstrd * __Cn;
+    __Cbstrd = __Cistrd * __Cm;
+
+    __transa = std::is_same<ArgTransA, Trans::NoTranspose>::value ? 'N' : 'T';
+    __transb = std::is_same<ArgTransB, Trans::NoTranspose>::value ? 'N' : 'T';
+  }
+
+  int invoke() {
+    if (__handle->enableDebug) {
+      std::cerr << "__nbatch:" << std::to_string(__nbatch)
+                << ", __ninter:" << std::to_string(__ninter)
+                << ", __Am:" << std::to_string(__Am)
+                << ", __An:" << std::to_string(__An)
+                << ", __alpha:" << std::to_string(__alpha)
+                << ", __beta:" << std::to_string(__beta) << std::endl;
+    }
+
+    if (!std::is_same<avt, double>::value ||
+        !std::is_same<bvt, double>::value ||
+        !std::is_same<cvt, double>::value ||
+        !std::is_same<ScalarType, double>::value) {
+      std::ostringstream os;
+      os << "KokkosBatched::Impl::BatchedArmplGemm only supports 'double' "
+            "scalar types."
+         << std::endl;
+      Kokkos::Impl::throw_runtime_exception(os.str());
+    }
+
+    if (__nbatch != 0) {
+      if (__ninter == 0 || __nbatch % __ninter) {
+        std::string msg =
+            "batch size must be evenly divisible by ninter. __nbatch: ";
+        msg += (std::to_string(__nbatch) +
+                ", __ninter: " + std::to_string(__ninter) + "\n");
+        Kokkos::abort(msg.c_str());
+      }
+
+      // Calculate internal batch size for interleaving
+      __nbatch /= __ninter;
+
+      // Allocate space for interleaving
+      //   __Adp and __Bdp are deleted in __run()
+      //   __Cdp is deleted in __repack_view()
+      __Adp = new avt[__Abstrd * __nbatch];
+      __Bdp = new bvt[__Bbstrd * __nbatch];
+      __Cdp = new cvt[__Cbstrd * __nbatch];
+
+      __unpack_views();
+      __run();
+      __repack_view();
```

Comment:
Can this support the cases of A, B, C are already interleaved? In these cases, ```__unpack_views()``` and ```__repack_views()``` are not necessary.

Reply:
Not yet; are you thinking that if the views have a SIMD vector scalar type, then we should skip the packing steps?

Reply:
Yes, it is what I am thinking. What are the costs of unpacking and repacking steps?
Will it take time to add this support? If yes, you can leave it to another PR.

Reply:
I will time the packing steps and get back to you but I suspect they are high (~50% of invoke). It won't take much time to add this.

Reply:
One more question: when ```niter = 1```, does this ArmPL interleaved batched gemm become the regular batched gemm?

ArmPL also provides the ```xgemm_batch``` (see this https://developer.arm.com/documentation/101004/2110/BLAS-Basic-Linear-Algebra-Subprograms/BLAS-Extensions/dgemm-batch?lang=en). I am not sure if these correspond to the interleaved batched gemm with ```niter = 1```.

Reply:
> One more question: when niter = 1, does this ArmPL interleaved batched gemm become the regular batched gemm?


That is my understanding.

> ArmPL also provides the xgemm_batch (see this https://developer.arm.com/documentation/101004/2110/BLAS-Basic-Linear-Algebra-Subprograms/BLAS-Extensions/dgemm-batch?lang=en). I am not sure if these correspond to the interleaved batched gemm with niter = 1.


We could look at a stack trace.

Reply:
> > One more question: when niter = 1, does this ArmPL interleaved batched gemm become the regular batched gemm?
> 
> That is my understanding.
> 
So, if ```niter = 1```, we might not need to allocate temporary arrays ```__Adp```, ```__Bdp```, ```__Cdp```, we can use the views ```__A```, ```__B```, ```__C```, instead.

Reply:
> I will time the packing steps and get back to you but I suspect they are high (~50% of invoke). It won't take much time to add this.

@vqd8a: On average for the input in these unit-tests, `__unpack_views` and `__repack_view` take 30% of the time and `__run` takes 70%. The total time used was the time spent in `__unpack_views`, `__repack_view`, and `__run`.